### PR TITLE
Fix for xcat-core/issues/233, rinv recieve incorrect data from BMC

### PR DESCRIPTION
**Pull request by whowutwut**
Commit 317d7cf contains the fix for Issue #233 , Running rinv results in the following messages coming back:
```
Error: Received incorrect data from BMC
```

The array contains uninitialized values that are not 0 or 1, so check if defined before processing the value in the array.  If not defined, return undef

The other commits are to fix some of the tab/space formatting in ipmi.pm.

### CODE REVIEW

**Comment by whowutwut** [comment]
This doesn't address all the issues in #233, will close for now and re-open if I do figure out how to fix the other issues.

**Comment by whowutwut** [comment]
@zet809  There are some issues with the ``rinv`` command on the Power S822LC machines that we have because of the problem with the FRU.  I think we need to open a ticket and get the FRU unit replaced ( @pdlun92 is working on that...  ) before we are able to successfully print out inventory information with ``rinv``.  Maybe if we can find a machine that has a working FRU, we can get onto there to do more devel work and possibly fix this in 2.11.1 if we can't get it fixed in 2.11.  

But for now, it seems to me that we are receiving data and if it's less than the chunksize, we print out the msg, but I think at the end of the data, the chunk received may be less than the size set... so are we printing out a msg that should not be printed out?  Can you review...? 

The code changes I made are in commit 317d7cf and 400711b.  The other 3 commits are fixing the space and formatting issues in ipmi.pm.  

If you don't agree. I can look further.....  

I think the issues @tingtli  brings up about the missing data can't be resolved until we fix the FRU unit...

**Comment by zet809** [comment]
@whowutwut I agree with you that not to output the error msg until we fix the FRU issue. 
Print out err msg for rinv or rvitials help us nothing but confusing users.

**Comment by whowutwut** [comment]
@zet809  if the changes are acceptable, please merge the pull request. Thanks!

**Comment by whowutwut** [comment]
@daniceexi @zet809 @pdlun92  There's a way to fix the FRU, I'm going to look into that, please don't merge this pull request until I see if we can fix the FRU information on the 822LC...

**Comment by cxhong** [comment]
regarding this issue,  the total fru size return back from fru inventory area info command (0x10) is 256.  for each fru id, we read FRU Data (command 0x11) from offset 0 to one chunk at a time, each chunk is 16 bytes, util we finish read all the 256 bytes.  But for some fru,  we asked for to read 16 bytes, but only have less than 16 bytes left,  I don't think this is incorrect data from BMC. so the code
````
       my @data = @{$rsp->{data}};
        if ($data[0] != $sessdata->{currfruchunk}) {
            add_fruhash($sessdata);
            xCAT::SvrUtils::sendmsg([1,"Received incorrect data from BMC"],$callback,$sessdata->{node},%allerrornodes);
            return;
        }
```````
I suggest to change to:
`````
       my @data = @{$rsp->{data}};     
        if ($data[0] < $sessdata->{currfruchunk}) {
            $chunk = $data[0];
            $sessdata->{currfrudone}=1;
        }
````````````
so we shrink chunk to only get the remainder data.

File: xCAT-server/lib/xcat/plugins/ipmi.pm [hunk 1]
```diff
@@ -545,98 +544,101 @@ sub on_bmc_connect {
...
+    }
+    else {
+        $rc = 1;
+        $text = "unsupported command $command $subcommand";
+    }
 
-	if($text) {
-		push(@output,$text);
-	}
+    if($debug) {
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/ipmi.pm]: There is a "return" in line 540, so this section won't work at all.

**Comment by zet809** [comment]
@cxhong The currfruchunk is calculated with the code below, so it must be fru error if returned data length not equal to currfrusize. 

if ($sessdata->{currfruoffset}+16  >= $sessdata->{currfrusize}) {
        $chunk =  $sessdata->{currfrusize}-$sessdata->{currfruoffset}; # shrink chunk to only get the remainder data
        $sessdata->{currfrudone}=1;
    }

**Comment by whowutwut** [comment]
After some discussion... For the 2.11 GA, will modify the ``rinv`` incorrect data message to be more desciptive and contain the FRU ID.  When we get access to a Power S822LC machine that has a good VPD info from MFG, will address this issue further and see if it's resolved.

New output will be as follows:
```
# rinv c910f05c33 bios
c910f05c33: Received incorrect data from BMC for FRU ID: 52
c910f05c33: Received incorrect data from BMC for FRU ID: 44
c910f05c33: Received incorrect data from BMC for FRU ID: 49
c910f05c33: Received incorrect data from BMC for FRU ID: 47
```

@zet809 , can you review the changes in the last two commits for this pull request 317d7cf08eb01f68be66d17bbf5f98888975ebed and 743960c5a81d8fc495ba6fdfa91a24595174f084 and merge the request if OK.  The others are formatting related.

**Comment by zet809** [comment]
The code looks good to me, I agree to merge it.

**Pull request merged**